service/sharddistributor/executorclient/clientimpl.go

```diff
@@ -0,0 +1,166 @@
+package executorclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/uber/cadence/client/sharddistributorexecutor"
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/service/sharddistributor/executorclient/syncgeneric"
+)
+
+type processorState int
+
+const (
+	processorStateStarting processorState = iota
+	processorStateStarted
+	processorStateStopping
+)
+
+type managedProcessor[SP ShardProcessor] struct {
+	processor SP
+	state     processorState
+}
+
+type executorImpl[SP ShardProcessor] struct {
+	logger                 log.Logger
+	shardDistributorClient sharddistributorexecutor.Client
+	shardProcessorFactory  ShardProcessorFactory[SP]
+	namespace              string
+	stopC                  chan struct{}
+	heartBeatInterval      time.Duration
+	managedProcessors      syncgeneric.Map[string, *managedProcessor[SP]]
+	executorID             string
+	timeSource             clock.TimeSource
+}
+
+func (e *executorImpl[SP]) Start(ctx context.Context) {
+	go e.heartbeatloop(ctx)
+}
+
+func (e *executorImpl[SP]) Stop() {
+	close(e.stopC)
```

Comment:
nit: This should have a WaitGroup to ensure that heartbeat goroutine is stopped once Stop is finished.